hugr/src/algorithm/const_fold.rs

```diff
@@ -19,6 +22,89 @@ use crate::{
     Hugr, HugrView, IncomingPort, Node, SimpleReplacement,
 };
 
+use super::VerifyLevel;
+
+#[derive(Error, Debug)]
+#[allow(missing_docs)]
+pub enum ConstFoldError {
+    #[error("Failed to verify {label} HUGR: {err}")]
+    VerifyError {
+        label: String,
+        #[source]
+        err: ValidationError,
+    },
+    #[error(transparent)]
+    SimpleReplaceError(#[from] SimpleReplacementError),
+}
+
+impl ConstFoldError {
+    fn verify_err(label: impl Into<String>, err: ValidationError) -> Self {
+        Self::VerifyError {
+            label: label.into(),
+            err,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Default)]
+/// TODO
+pub struct ConstFoldConfig {
+    verify: VerifyLevel,
+}
+
+impl ConstFoldConfig {
+    /// TODO
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// TODO
+    pub fn with_verify(mut self, verify: VerifyLevel) -> Self {
+        self.verify = verify;
+        self
+    }
+
+    fn verify_impl(
+        &self,
+        label: &str,
+        h: &impl HugrView,
+        reg: &ExtensionRegistry,
+    ) -> Result<(), ConstFoldError> {
+        match self.verify {
+            VerifyLevel::None => Ok(()),
+            VerifyLevel::WithoutExtensions => h.validate_no_extensions(reg),
+            VerifyLevel::WithExtensions => h.validate(reg),
+        }
+        .map_err(|err| ConstFoldError::verify_err(label, err))
+    }
+
+    /// TODO
+    pub fn run(&self, h: &mut impl HugrMut, reg: &ExtensionRegistry) -> Result<(), ConstFoldError> {
```

Comment:
not sure I like having a `run` method on something with the suffix `Config`. There's the beginnings of a general "algorithm" interface here (configuration, transformation, verfification). Is it worth deferring some of these changes in to a generic follow up PR?